### PR TITLE
Fix T8 coefficient in matrix exponential

### LINKED ISSUE #196592: matrix_exp T8 approximation uses an incorrect x7 coefficient
### 🐛 Describe the bug

`torch.linalg.matrix_exp` / `torch.matrix_exp` uses an incorrect coefficient
in the degree-8 Taylor approximation implemented by `compute_T8`.

For a small 2D rotation, this causes a reproducible numerical error when a
singleton matrix selects the T8 path.

A minimal reproducer is:

```python
import math
import torch

for dtype, theta in [(torch.float32, 0.5), (torch.float64, 0.04)]:
    a = torch.tensor([[0.0, -theta], [theta, 0.0]], dtype=dtype)
    t = a[1, 0].item()

    expected = torch.tensor(
        [
            [math.cos(t), -math.sin(t)],
            [math.sin(t),  math.cos(t)],
        ],
        dtype=torch.float64,
    )

    for count in (1, 2):
        result = torch.matrix_exp(a.repeat(count, 1, 1))[0]
        error = (result.double() - expected).abs().max().item()
        print(dtype, count, error)
```

On macOS ARM64 CPU with PyTorch 2.14.0:
```text
torch.float32 1 2.4903838578005377e-05
torch.float32 2 1.0736774491704182e-07
torch.float64 1 8.331875467337824e-11
torch.float64 2 2.220446049250313e-16
```

The analytic reference is the closed-form exponential of
```text
A = [[0, -theta],
     [theta, 0]]
```

which is the 2D rotation matrix
```text
exp(A) = [[cos(theta), -sin(theta)],
          [sin(theta),  cos(theta)]].
```

### Root cause

`compute_T8` currently contains:
```cpp
constexpr scalar_t x7 = (89. - sqrt_177) / (5040. * x3);
```

However, the construction cited by the implementation requires:
```cpp
constexpr scalar_t x7 = (89. - sqrt_177) / (5040. * x3 * x3);
```
with `x3 = 2 / 3`.

Bader, Blanes, and Casas, "Computing the Matrix Exponential with an Optimized Taylor Polynomial Approximation", Mathematics 2019, 7, 1174, Eq. (13).

This can also be verified independently by expanding the factorized polynomial.
With the current coefficient, the polynomial first differs from the degree-8 Taylor series at the `A^5` term. With the additional factor of `x3` in the denominator, all coefficients through degree 8 are exactly `1/k!`.

### Batch-dependent behavior

The difference between batch size 1 and batch size 2 is explained by the current
implementation.

For singleton inputs, `mexp_impl` selects among T1/T2/T4/T8/T12 according to the matrix norm. 
A flattened batch size greater than 1 instead takes the T18 scaling-and-squaring path.

Therefore duplicating the same matrix into a batch avoids the faulty T8 approximation; 
it is not an accuracy improvement caused by batching itself.

For the examples above, T8 is selected for the singleton matrices:

- float32: `0.05116619 < ||A||_1 <= 0.58005246`
- float64: `0.00033971688 < ||A||_1 <= 0.04991229`

### Backward

The reverse-mode differential evaluates a block matrix exponential containing the input and cotangent, so it is also affected when that block matrix selects T8.

For the float32 example above with a cotangent G = 0.01 * I, the native VJP has a max absolute error of approximately 2.47e-6. With the corrected T8 coefficient, the same diagnostic has error below 1e-9.

This does not mean every backward call corresponding to an inaccurate forward is inaccurate: the cotangent can change the norm of the block matrix and therefore cause a different approximation degree to be selected.

### Proposed fix

The missing factor is a one-line correction:

```cpp
- constexpr scalar_t x7 = (89. - sqrt_177) / (5040. * x3);
+ constexpr scalar_t x7 = (89. - sqrt_177) / (5040. * x3 * x3);
```

I have a minimal patch and focused forward/backward regression tests ready locally. I built PyTorch from source with the correction and verified that the reproducer, the new regressions, the existing `matrix_exp` tests, and `spin quicklint` pass on my machine.

No PR has been opened yet; I can submit it once this issue is considered actionable.

### AI assistance

I used AI coding assistants during the investigation to help inspect source paths, prepare diagnostic scripts, and edit parts of this report and the regression tests. I reviewed and verified the reproducer, mathematical derivation, patch, and test results myself.

### Versions
```text
PyTorch version: 2.14.0
Is debug build: False
CUDA used to build PyTorch: None
ROCm SDK used to build PyTorch: N/A
HIP used to build PyTorch: N/A

OS: macOS 26.6.2 (arm64)
GCC version: Could not collect
Clang version: 21.0.0 (clang-2100.1.1.101)
CMake version: version 4.4.3
Libc version: N/A

Python version: 3.12.13 (main, Jun  2 2026, 22:42:28) [Clang 22.1.3 ] (64-bit runtime)
Python platform: macOS-26.6.2-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Apple M5 Pro

Versions of relevant libraries:
[pip3] Could not collect
[conda] Could not collect
```

cc @jianyuh @nikitaved @mruberry @walterddr @Lezcano

### PR DESCRIPTION
**Pull request by Concode0**
## Issue

Fixes #196592

## Summary

Correct the missing `x3` factor in the `compute_T8` matrix exponential coefficient and add focused regression tests for the affected forward and backward paths.

The tests cover singleton inputs that select the T8 path, along with batched inputs for comparison.

Local validation:
- `python test/test_linalg.py -k matrix_exp_small_rotation`
- `python test/test_linalg.py -k matrix_exp`
- `spin quicklint`

## Checklist

- [x] Added/updated tests
- [ ] Updated documentation (not applicable)
- [ ] Included benchmark results (not applicable; no performance change)

## BC-breaking?

No.

### AI assistance

I used AI coding assistants during the investigation to help inspect source paths, prepare diagnostic scripts, and edit parts of the regression tests and write-up. I reviewed and verified the patch, mathematical derivation, and test results myself.

cc @jianyuh @nikitaved @mruberry @walterddr @Lezcano

### CODE REVIEW

File: test/test_linalg.py [hunk 1]
```diff
@@ -8241,6 +8241,44 @@ def run_test(*n):
...
+    @skipCPUIfNoLapack
+    @dtypes(torch.float, torch.double, torch.cfloat, torch.cdouble)
+    def test_linalg_matrix_exp_small_rotation(self, device, dtype):
+        single = dtype in (torch.float, torch.cfloat)
+        theta = 0.5 if single else 0.04
+        atol = 3e-7 if single else 2e-15
+        a = torch.tensor([[0, -theta], [theta, 0]], device=device, dtype=dtype)
+        # Form the reference in double precision from the represented input.
+        t = a[1, 0].real.cpu().double()
+        c, s = t.cos(), t.sin()
```
Comment (nikitaved) [inline test/test_linalg.py]: theta is already defined, no need to extract it from a tensor residing on a gpu. And you can wrap it into a Tensor, if needed. Same for the following test.

File: test/test_linalg.py [hunk 2]
```diff
@@ -8241,6 +8241,44 @@ def run_test(*n):
...
+        expected = torch.stack((torch.stack((c, -s)), torch.stack((s, c))))
+        expected = expected.to(device=device, dtype=dtype)
+        for batch_shape in ((), (1,), (2,), (1, 2)):
+            x = a.expand(*batch_shape, 2, 2).contiguous()
```
Comment (nikitaved) [inline test/test_linalg.py]: Does it work if `contiguous` is dropped?

Comment (Concode0) [reply test/test_linalg.py]: Yes, it works without contiguous; I dropped it from both tests.

File: test/test_linalg.py (hunk 1 above)
Comment (Concode0) [reply test/test_linalg.py]: Updated both tests to use theta directly. Thanks!

File: test/test_linalg.py [hunk 3]
```diff
@@ -8241,6 +8241,46 @@ def run_test(*n):
...
+        )
+        for batch_shape in ((), (1,), (2,), (1, 2)):
+            x = a.expand(*batch_shape, 2, 2)
+            self.assertEqual(torch.linalg.matrix_exp(x), expected.expand_as(x),
+                             atol=atol, rtol=0)
+
+    @skipCUDAIfNoMagmaAndNoLinalgsolver
+    @skipCPUIfNoLapack
+    @dtypes(torch.float, torch.double, torch.cfloat, torch.cdouble)
+    def test_linalg_matrix_exp_small_rotation_backward(self, device, dtype):
```
Comment (nikitaved) [inline test/test_linalg.py]: nit: I think these two tests can be merged -- but up to you. Please, add a link to the issue though. BTW, maybe existing tests have similar cases and can just be expanded?

Comment (Concode0) [reply test/test_linalg.py]: Thanks. The existing analytic/Taylor tests cover similar degree ranges, but their tolerances are too loose to catch this regression. I'll keep the focused regression case, but merge the forward/backward checks, and add a link to the issue.

Comment (nikitaved) [reply test/test_linalg.py]: Sounds good!

**Review by lezcano — APPROVED** [review]

**Comment by nikitaved** [comment]
Thank you, @Concode0 , for the fix!
